python/llm/example/GPU/Speculative-Decoding/Eagle/evaluation/gen_ea_answer_llama2chat.py

```diff
@@ -0,0 +1,500 @@
+        )
+        if tokenizer.eos_token_id in input_ids[0, input_len:].tolist():
+            break
+        if new_token > 1024:
+            break
+        if input_ids.shape[1] > 1960:
+            break
+    return input_ids, new_token, idx
+
+
+def run_eval(
+        base_model_path,
+        ea_model_path,
+        model_id,
+        question_file,
+        question_begin,
+        question_end,
+        answer_file,
+        max_new_token,
+        num_choices,
+        num_gpus_per_model,
+        num_gpus_total,
+        max_gpu_memory,
+        temperature,
+        tree_choices,
+        enable_ipex_llm,
+):
+    questions = load_questions(question_file, question_begin, question_end)
+    shuffled_ids = [q["question_id"] for q in questions]
+
+    # Split the question file into `num_gpus` files
+    assert num_gpus_total % num_gpus_per_model == 0
+    use_ray = num_gpus_total // num_gpus_per_model > 1
+
+    if use_ray:
+        get_answers_func = ray.remote(num_gpus=num_gpus_per_model)(
+            get_model_answers
+        ).remote
+    else:
+        get_answers_func = get_model_answers
+
+    chunk_size = len(questions) // (num_gpus_total // num_gpus_per_model)  # // 2
+    ans_handles = []
+    for i in range(0, len(questions), chunk_size):
+        ans_handles.append(
+            get_answers_func(
+                base_model_path,
+                ea_model_path,
+                model_id,
+                questions[i: i + chunk_size],
+                answer_file,
+                max_new_token,
+                num_choices,
+                num_gpus_per_model,
+                max_gpu_memory,
+                temperature,
+                tree_choices,
+                enable_ipex_llm,
+            )
+        )
+
+    if use_ray:
+        ray.get(ans_handles)
+
+
+@torch.inference_mode()
+def get_model_answers(
+        base_model_path,
+        ea_model_path,
+        model_id,
+        questions,
+        answer_file,
+        max_new_token,
+        num_choices,
+        num_gpus_per_model,
+        max_gpu_memory,
+        temperature,
+        tree_choices,
+        enable_ipex_llm,
+):
+    try:
+        model = EaModel.from_pretrained(
+            base_model_path=base_model_path,
+            ea_model_path=ea_model_path,
+            #torch_dtype=torch.float16,
+            torch_dtype=torch.float32,
+            low_cpu_mem_usage=True,
+            # load_in_8bit=True,
+            device_map="auto"
+        )
+    except ValueError:
+        print("Using sequential device_map.")
+        model = EaModel.from_pretrained(
+            base_model_path=base_model_path,
+            ea_model_path=ea_model_path,
+            #torch_dtype=torch.float16,
+            torch_dtype=torch.float32,
+            low_cpu_mem_usage=True,
+            # load_in_8bit=True,
+            device_map="sequential"
+        )
+    if enable_ipex_llm:
+        model = optimize_model(model, optimize_llm=False)
+    model.to("xpu")
```

Comment:
In my understanding, from line 197 to line 220 is the only change for ipex-llm ? maybe we can add a comment here to emphasize that the code is replated to ipex-llm, so that it is easier to distinguish the original code from the modified code.

Reply:
strictly speaking, only line 219 is required to enable ipex-llm. I added a comment as suggested in each of the GPU and CPU files.